dimod/binary_quadratic_model/model.py

```diff
@@ -0,0 +1,1099 @@
+        adj = self.adj
+        linear = self.linear
+
+        if value not in self.vartype.value:
+            raise ValueError("expected value to be in {}, received {} instead".format(self.vartype.value, value))
+
+        removed_interactions = []
+        for u in adj[v]:
+            self.add_variable(u, value * adj[v][u])
+            removed_interactions.append((u, v))
+        self.remove_interactions_from(removed_interactions)
+
+        self.add_offset(value * linear[v])
+        self.remove_variable(v)
+
+    def flip_variable(self, v):
+        """Flips a single variable v.
+
+        Args:
+            v (variable):
+                A variable in the binary quadratic model.
+
+        Notes:
+            If v is not in the binary quadratic model then it is ignored.
+
+        Examples:
+            >>> bqm = dimod.BinaryQuadraticModel({'a': -1}, {}, 0.0, dimod.SPIN)
+            >>> original = bqm.copy()
+            >>> bqm.flip_variable('a')
+            >>> bqm.energy({'a': -1}) == bqm.energy({'a': 1})
+            True
+
+        """
+        adj = self.adj
+        linear = self.linear
+        quadratic = self.quadratic
+
+        if v not in adj:
+            return
+
+        if self.vartype is Vartype.SPIN:
+            # in this case we just multiply by -1
+            linear[v] *= -1.
+            for u in adj[v]:
+                adj[v][u] *= -1.
+                adj[u][v] *= -1.
+
+                if (u, v) in quadratic:
+                    quadratic[(u, v)] *= -1.
+                elif (v, u) in quadratic:
+                    quadratic[(v, u)] *= -1.
+                else:
+                    raise RuntimeError("quadratic is missing an interaction")
+
+        elif self.vartype is Vartype.BINARY:
+            self.offset += linear[v]
+            linear[v] *= -1
+
+            for u in adj[v]:
+                bias = adj[v][u]
+
+                adj[v][u] *= -1.
+                adj[u][v] *= -1.
+
+                linear[u] += bias
+
+                if (u, v) in quadratic:
+                    quadratic[(u, v)] *= -1.
+                elif (v, u) in quadratic:
+                    quadratic[(v, u)] *= -1.
+                else:
+                    raise RuntimeError("quadratic is missing an interaction")
+
+        else:
+            raise RuntimeError("Unexpected vartype")
+
+        try:
+            self._counterpart.flip_variable(v)
+        except AttributeError:
+            pass
+
+    def update(self, bqm):
+        """Update with the values from another binary quadratic model.
+
+        Args:
+            bqm (:class:`.BinaryQuadraticModel`):
+                A binary quadratic model. All of the biases are added to self.
+
+        """
+        self.add_variables_from(bqm.linear, vartype=bqm.vartype)
+        self.add_interactions_from(bqm.quadratic, vartype=bqm.vartype)
+        self.add_offset(bqm.offset)
+
+    def contract_variables(self, u, v):
+        """Asserts u, v are the same variable.
```

Comment:
Not sure about the wording of "asserts" here.

Reply:
Try enforces